### PR TITLE
Add a compile-time switch for cgroup unified hierarchy handling

### PR DESCRIPTION
**Pull request by zonque**
This PR carries a fix for `cg_is_unified_systemd_controller_wanted()` and a new configure switch to entirely disable all unified cgroup usage, even on kernels that support it. That's important for shipping systemd binaries that should be forward-compatible to systems with newer kernels.

### CODE REVIEW

**Comment by evverx** [comment]
> switch to entirely disable all unified cgroup usage

@zonque , why not to use `systemd.unified_cgroup_hierarchy=no systemd.legacy_systemd_cgroup_controller=yes`?

File: src/basic/cgroup-util.c [hunk 1]
```diff
@@ -2401,7 +2401,7 @@ bool cg_is_unified_systemd_controller_wanted(void) {
         /* If the unified hierarchy is requested in full, no need to
          * bother with this. */
         if (cg_is_unified_wanted())
-                return 0;
+                return true;
```
Comment (keszybz) [inline src/basic/cgroup-util.c]: No, no, please read the comment above: "no need to bother with this".

Comment (zonque) [reply src/basic/cgroup-util.c]: Ah, confusion on my side. Sure, if we have a fully unified system, the systemd controller needs no special treatment. Will turn it into `false` though.

**Comment by zonque** [comment]
> I don't get the reason for this patch... Currently the default in systemd is not to use the unified hierarchy, even if available.

Yes it does. #3965 introduced a "hybrid mode" that will use the unified hierarchy in parallel to the legacy hierarchies. That way, we can make use of the nifty features for process tracking in v2 (unified), while still using the controllers of v1 (legacy).

A potential problem with that is that systemd binaries will lead to different run-time behavior, depending on whether the kernel supports v2 or not. The purpose of the patch is to have a compile-time switch that allows distros to build systemd without v2, regardless of whether the kernel supports it or not.

Makes sense?

**Comment by zonque** [comment]
> switch to entirely disable all unified cgroup usage
> 
> @zonque , why not to use systemd.unified_cgroup_hierarchy=no systemd.legacy_systemd_cgroup_controller=yes?

That's a run-time switch. What that patch adds is a compile-time switch that does the same thing.

**Comment by evverx** [comment]
> What that patch adds is a compile-time switch that does the same thing.

@zonque , I'm not sure. Seems like this PR totally disables a run-time switch.
I mean:

```
./conifure --disable_unified_cgroup_hierarchy
```

How can I enable the unified hierarchy at run-time?

**Comment by zonque** [comment]
> How can I enable the unified hierarchy at run-time?

You cannot. That's the whole point.

**Comment by evverx** [comment]
@zonque , oh, I understand.

Also, this PR breaks https://github.com/systemd/systemd/pull/4310
i.e. this combination
- host: `systemd-with-unified-enabled-v232`
- container: `systemd-with-unified-disabled-v232`

will not work

**Comment by zonque** [comment]
> Also, this PR breaks #4310
> i.e. this combination
> 
> ```
> host: systemd-with-unified-enabled-v232
> container: systemd-with-unified-disabled-v232
> ```
> 
> will not work

True. That setting is intended for systemd on the host only. For containers to be able to adopt to whatever the host offers them, they need to be built with v2 enabled, ie, in auto-detect mode.

**Comment by htejun** [comment]
I can see why we would want a compile time switch to select a specific default mode but disabling v2 support completely doesn't make sense to me, especially given that it would close off things like auto-detecting v2 mode when run in container and so on.

**Comment by keszybz** [comment]
> #3965 introduced a "hybrid mode" that will use the unified hierarchy in parallel to the legacy hierarchies

The hybrid mode is mostly a systemd-internal thing, at least it was supposed to be so. Life is of course slightly more complicated…

We normally (as in Fedora, and other distros) use a single build for both host and containers. A compile-time switch would break use of that binary in containers on a host that uses unified hierarchy. Even if we don't want to do that currently by default, at some point we will, and we want the systemds installed in various containers to not break.

> I can see why we would want a compile time switch to select a specific default mode

Ack, we should make it possible to choose systemd hierarchy independent of the kernel support. So `--cgroup-hierarchy=legacy|unified|hybrid`, with current default of hybrid would be useful.

**Comment by poettering** [comment]
So, yupp, I'd vote too for making this a compile time switch to pick the default mode, but still permit any other mode to be specified on the kernel cmdline. I mean, I think we should always build in support for all three modes anyway, as systemd running in a container should really make the best of what it gets, regardless what it is configured for. Hence, in order to determine which mode to use, we should maybe go for the following logic:
1. if systemd starts and finds cgroups already mounted in one of the three modes, use it, do not alter anything. This is relevant for two cases: so that PID 1 on the host reuses the cgroup setup prepared by the initrd, even if the initrd and the host systemd are different versions. And secondly, so that PID 1 in a container makes use of whichever of the three modes was set up by the container manager.
2. If nothing is mounted yet, look for the kernel cmdline setting: systemd.unified_cgroup_hierarchy=yes|no|hybrid. If set, it counts and is what is set up (except if unified is requested this way but the kernel doesn't support it at all, in that case use legeacy)
3. Otherwise, use whatever was specified on the configure command line, using --with-unified-cgroup-hierarchy=yes|no|hybrid (and of course, downgrade to legacy if kernel doesn't do unified)
4. If that wasn't set, and cpu+memory+io are available as unified controllers (and only those, the other ones are secondary I'd say and don't really matter too much), use full unified mode
5. Otherwise, if any of cpu/memory/io is missing but unified otherwise available, use hybrid mode
6. Otherwise, use legacy mode

Of course, we could also call the option systemd.cgroup_hierarchy=legacy|unified|hybrid (and matching --with-cgroup-hierarchy on configure cmdline), but the "systemd.unified_cgroup_hierarchy" kernel cmdline one already exists, hence I figure we should stick to it.

With the above scheme for current vanilla kernels and no explicit config we'd default to "hybrid" mode. Distros which want to support current Docker and want stable kernel-level cgroup APIs should however use the configure switch to change the default to "legacy" mode. RHEL would certainly be in this camp. Fedora should at least during the Rawhide cycles not configure anything explicitly, so that "hybrid" is the default for them too. When freezing they might want to change to "legacy" too however, or maybe not, they'd need to figure that out.

**Comment by keszybz** [comment]
Sounds good, with one exception:

> 1. If that wasn't set, and cpu+memory+io are available as unified controllers (and only those, the other ones are secondary I'd say and don't really matter too much), use full unified mode

I think that's too much magic: currently a patched kernel is required, so
this isn't getting much testing. At some point in the future cpu controller
will be exposed in the upstream kernel, so this mode might get switched on
after a kernel upgrade. This sounds dangerous — we should first wait for
upstream support in it's final form, and only then consider switching the
default to v2. I also think we should do with loud&clear announcement
(NEWS, etc).

> 1. Otherwise, if any of cpu/memory/io is missing but unified otherwise available, use hybrid mode

I'd got for this as the default.

> 1. Otherwise, use legacy mode

**Comment by evverx** [comment]
> except if unified is requested this way but the kernel doesn't support it at all, in that case use legeacy

Oh, this makes sense

``` sh
# 4.4.0-43-generic
# systemd.unified_cgroup_hierarchy=yes

[!!!!!!] Failed to mount API filesystems, freezing.
[    1.818148] systemd[1]: Freezing execution.
```

**Comment by zonque** [comment]
Okay, let me redo this. Thanks for all the comments!

**Pull request closed without merging**